src/polyfem/solver/AdjointNLProblem.cpp

```diff
@@ -290,6 +486,26 @@ namespace polyfem::solver
 				need_rebuild_basis = true;
 		}
 
+		// Apply slim to all states on a frequency
+		if (need_rebuild_basis && enable_slim && curr_x.size() > 0)
```

Comment:
Why not remove the slim from "is_step_valid" and only keep it here? If solution_changed throws an error, the line search would reduce step size and retry.